### ISSUE TITLE
JSON-RPC client doesn't support asynchronous non-blocking messages

### ISSUE THREAD

**Issue by tymonx** [cinemast/libjson-rpc-cpp#96]
JSON-RPC client doesn't support asynchronous non-blocking messages.

src/jsonrpccpp/client/connectors/httpclient.cpp should have possibility to use curl_multi_perform: http://curl.haxx.se/libcurl/c/curl_multi_perform.html

**Comment by cinemast** [comment]
Hi tymonx!

I don't like the idea of that. What would be the advantage of switching to multi_perform?
Any application can achieve this by calling JSON-RPC methods from threads. 
The CURL handle of HttpClient is on the stack and therefore thread safe.

Except for the first call to HttpClient (which I fixed in the latest commit on dev: 2ff44b23925d08d81302462ddb623c99f6281548

**Comment by cinemast** [comment]
If you can convince me of an advantage, I will reconsider the wontfix.

**Comment by tymonx** [comment]
You don't take advantage using ID field from JSON-RPC specification. In synchronous mode when client talking to server 1-to-1 with fixed IP and PORT, you shouldn't send another message (e.g in thread) during that operation. Response with ID=1 may come to ID=2 requester and vice versa. Network is unpredictable. This issue drop throughput performance in that synchronous scenario and is dependent on the method time execution on the server side. Also during send call, client must wait for response.

For maximum throughput, the best solution may be using non-blocking asynchronous messages:
* put client requests to queue with callback or better with std::function
* get requests from queue in single thread and asynchronous sending to server
* wait for messages and compare with ID (it may return with different order!)
* execute callback, function to user with received response

I suggest implement send method in 2 ways:
```C
response = send(request) // Old-way, synchronous call
send(request, std::function, optional=timeout) // Non-blocking asynchronous send, C++11 function: lambda, class, function
```

**Comment by cinemast** [comment]
Hi tymonx!

The thing with the request/response confusion is not true. CURL handles it pefectly with multiple threads, as those are separate TCP connections. So everything is predictable in any case.

However the performance argument is valid. I do not take advantage of the IDs field. But I am fine with that, as it keeps things clean and simple.

I think the "performance-gain" is not worth in changing the whole calling paradigma of this framework. Aside that, if performance is really that critical for your usecase, no one prevents you from implementing your own stateful `IClientConnector` sub-class e.g. `AsyncHttpClient`. 

`std::function` would also drop non C++-11 compatibility.

So if you write such a class, you can raise a pull-request and I will review and maybe merge it.
I would appreciate such a contribution, just not as a replacement for the existing solution. 

Thank you for your input :+1:

**Comment by tymonx** [comment]
About that:

> The thing with the request/response confusion is not true. CURL handles it pefectly with multiple threads, as those are separate TCP connections. So everything is predictable in any case.

Yes, I agree that is not true, because you always create and close socket per one RPC message and this may be used in thread safe sending operation because every time source port from client is new. This solution also has negative side, drop performance, dynamic allocation, request to reserve and free buffer every time from OS. In my opinion network connection between client-server should be established and here show up this issue. In my case, I have main application and 0-N agents, both have client/server RPC. With your open/close solution per message I cannot recognize from who message come. This force me to use new port for every server on main application side. For large network system this is a waste of resources.

I apologize that I wrote that shortly without whole context, because I have done some modification in your library.

Using C++11 features can boost your code (e.g. move semantics).

**Comment by cinemast** [comment]
I will close this issue for now. You can raise a PR if you want to add this feature later. About C++11: I did not yet have time to take a deeper look into this.

Greetings

**Comment by vjpr** [comment]
Why does the homepage say:

> Full JSON-RPC 2.0 & 1.0 Client and Server Support.

...when it does not make use of ids?

I am working in an embedded environment and use unix pipes over ARM semihosting. This means I have two simplex channels to communicate on. The order messages are sent and received are an issue.

I think there should be a mention of this as a caveat on the homepage.

**Comment by cinemast** [comment]
Could you be a little bit more specific to which extend ids are not enabled by this framework?

**Comment by vjpr** [comment]
The `Client::CallMethod` impl is blocking. As @tymonx mentioned it would be great if there was a central dispatcher to allow out-of-order messages over a single connection, instead of having to use threads/tcp connections.

However, now that I have started implementing the library it is working great. I will just have to find a way to block in the `SendRPCMessage`.

**Comment by cauthu** [comment]
> Full JSON-RPC 2.0 & 1.0 Client and Server Support.

That tricked me into spending time looking around for how to send an RPC from server --> client, which is part of the 1.0 spec. After a while looking around the code and not finding anything, I looked at the issues and found issue #138 where you say:

> It is currently not directly supported. What you can do is host on each peer a JSON-RPC Server and Client. This way they can communicate in both directions.

I would suggest rephrasing the "Full JSON-RPC 2.0 & 1.0 support" claim it so people don't lose time mistakenly trying to do something the library doesn't support.

With that said, it seems a nice library.

**Comment by VishwanathPatil** [comment]
I encountered similar issue, that is CallMethod blocks, I want to unblock is there a way ?
